spec/fastsync/fastsync_concurrency_with_peers.tla

```diff
@@ -0,0 +1,543 @@
+      to the buffer of the current peer
+*)
+SendPeerMsgs ==
+    /\ ~Empty(peerIncoming[peerTurn])
+    /\ Ready(peerOutgoing[peerTurn])
+    /\ \/ /\ GetMessage(peerIncoming[peerTurn]) = "bcStatusRequest"
+          /\ peerOutgoing' \in PeerSend(peerOutgoing, peerTurn, {"bcStatusResponse"})
+       \/ /\ GetMessage(peerIncoming[peerTurn]) = "scBlockRequest"
+          /\ peerOutgoing' \in PeerSend(peerOutgoing, peerTurn, {"bcBlockResponse"})
+       \/ /\ peerOutgoing' \in PeerSend(peerOutgoing, peerTurn, peerMsgs)
+    /\ peerIncoming' = [peerIncoming EXCEPT ![peerTurn].queue = Tail(peerIncoming[peerTurn].queue)]  
+    /\ peerTurn' \in PeerIDs
+    /\ UNCHANGED <<receiveDemux, demuxSend, demuxProcess, processDemux, demuxSchedule, scheduleDemux>>
+
+(* 
+ReceiveRoutine --
+    either handle the message sent by the current peer,
+    or do nothing 
+*)
+ReceiveRoutine ==
+    \* handle the incoming message from the peer
+    \/ HandlePeerReceive
+    \* do nothing
+    \/ /\ UNCHANGED buffers
+       /\ UNCHANGED peer
+    
+(*
+DemuxRoutine --
+    either handle the incoming messages on the receiveDemux, processDemux, or scheduleDemux buffers,
+    or broadcast a status request to all peers,
+    or send a "rProcessBlock" to the process routine,
+    or do nothing
+*)    
+DemuxRoutine ==
+    \* handle the incoming messages on either receiveDemux, processDemux or scheduleDemux
+    \/ HandleReceiveDemux
+    \/ HandleProcessDemux
+    \/ HandleScheduleDemux
+       
+    \* broadcast a status request
+    \/ BroadcastStatusRequest 
+       
+    \* send a "rProcessBlock" to the process routine
+    \/ SendRProcessBlock
+       
+    \* otherwise, do nothing       
+    \/ /\ UNCHANGED buffers
+       /\ UNCHANGED peer
+
+(* 
+ProcessRoutine --
+    either handle the message sent by the demux routine on the demuxProcess buffer,
+    or do nothing 
+*)
+ProcessRoutine ==
+    \* handle the incoming messages on demuxProcess
+    \/ HandleDemuxProcess
+    \* do nothing
+    \/ /\ UNCHANGED buffers
+       /\ UNCHANGED peer
+
+(* 
+ScheduleRoutine --
+    either handle the message sent by the demux routine on the demuxSchedule buffer,
+    or do nothing 
+*)
+ScheduleRoutine ==
+    \* handle the incoming message on demuxSchedule
+    \/ HandleDemuxSchedule
+    \* do nothing
+    \/ /\ UNCHANGED buffers
+       /\ UNCHANGED peer
+
+(* 
+SendRoutine --
+    either handle the message sent by the demux routine on the demuxSend buffer,
+    or do nothing 
+*)
+SendRoutine ==
+    \* handle the incoming message on demuxSend
+    \/ HandleDemuxSend
+    \* do nothing
+    \/ /\ UNCHANGED buffers
+       /\ UNCHANGED peer
+      
+(*
+Peer --
+    either non-deterministically send peer messages to the receive routine
+    or do noting
+*)       
+Peer ==
+    \* send peer messages
+    \/ SendPeerMsgs
+    \* do nothing 
+    \/ /\ UNCHANGED buffers
+       /\ UNCHANGED peer
+
+Buffer ==
```

Comment:
The name is kind of misleading. It captures the behavior of multiple buffers, not of one.